src/convertor_tests.rs

```diff
@@ -246,6 +248,123 @@ fn read_collection() -> Result<Vec<RevlogEntry>> {
     Ok(revlogs)
 }
 
+#[test]
+fn extract_simulator_config_from_revlog() {
+    let revlogs = read_collection().unwrap();
+    let first_rating_count = revlogs
+        .iter()
+        .filter(|r| r.review_kind == RevlogReviewKind::Learning && r.ease_factor == 0)
+        .counts_by(|r| r.button_chosen);
+    let first_rating_prob = first_rating_count
+        .iter()
+        .map(|(button_chosen, count)| {
+            (
+                *button_chosen,
+                *count as f32 / first_rating_count.values().sum::<usize>() as f32,
+            )
+        })
+        .collect::<HashMap<_, _>>();
```

Comment:
Because we know there are only 4 possible values here, it would be simpler to collect directly into an array here. You can use this helper which I'm happy to license as BSD-3 here: https://github.com/ankitects/anki/blob/59759b468f03da47f33ea2867c0a0f773e8fcd02/rslib/src/scheduler/fsrs/try_collect.rs#L11

Like this:

let first_rating_prob: [f32; 4] = ...try_collect().unwrap()

Reply:
invalid_input!() would need to be changed to return Err(FSRSError::InvalidInput)